pkg/detectors/dockerhubv2/dockerhubv2.go

```diff
@@ -0,0 +1,179 @@
+
+var (
+	// Can use email or username for login.
+	usernamePat = regexp.MustCompile(`(?im)(?:user|usr|-u|id)\S{0,40}?[:=\s]{1,3}[ '"=]?([a-zA-Z0-9]{4,40})\b`)
+	emailPat    = regexp.MustCompile(`(` + common.EmailPattern + `)`)
+
+	// Can use password or personal access token (PAT) for login, but this scanner will only check for PATs.
+	accessTokenPat = regexp.MustCompile(detectors.PrefixRegex([]string{"docker"}) + `\b([a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12})\b`)
+)
+
+// Keywords are used for efficiently pre-filtering chunks.
+// Use identifiers in the secret preferably, or the provider name.
+func (s Scanner) Keywords() []string {
+	return []string{"docker"}
+}
+
+// FromData will find and optionally verify Dockerhub secrets in a given set of bytes.
+func (s Scanner) FromData(ctx context.Context, verify bool, data []byte) (results []detectors.Result, err error) {
+	dataStr := string(data)
+
+	// Deduplicate results.
+	tokens := make(map[string]struct{})
+	for _, matches := range accessTokenPat.FindAllStringSubmatch(dataStr, -1) {
+		tokens[matches[1]] = struct{}{}
+	}
+	if len(tokens) == 0 {
+		return
+	}
+	usernames := make(map[string]struct{})
+	for _, matches := range usernamePat.FindAllStringSubmatch(dataStr, -1) {
+		usernames[matches[1]] = struct{}{}
+	}
+	for _, matches := range emailPat.FindAllStringSubmatch(dataStr, -1) {
+		usernames[matches[1]] = struct{}{}
+	}
+
+	// Process results.
+	for token := range tokens {
+		s1 := detectors.Result{
+			DetectorType: s.Type(),
+			Raw:          []byte(token),
+		}
+
+		for username := range usernames {
+			s1.RawV2 = []byte(fmt.Sprintf("%s:%s", username, token))
+
+			if verify {
+				if s.client == nil {
+					s.client = common.SaneHttpClient()
+				}
+
+				isVerified, extraData, verificationErr := s.verifyMatch(ctx, username, token)
+				s1.Verified = isVerified
+				s1.ExtraData = extraData
+				s1.SetVerificationError(verificationErr)
+			}
+
+			results = append(results, s1)
+
+			if s1.Verified {
+				break
+			}
+		}
+
+		// PAT matches without usernames cannot be verified but might still be useful.
+		if len(usernames) == 0 {
+			results = append(results, s1)
+		}
+	}
+	return
+}
+
+func (s Scanner) verifyMatch(ctx context.Context, username string, password string) (bool, map[string]string, error) {
+	payload := strings.NewReader(fmt.Sprintf(`{"username": "%s", "password": "%s"}`, username, password))
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, "https://hub.docker.com/v2/users/login", payload)
+	if err != nil {
+		return false, nil, err
+	}
+
+	req.Header.Add("Content-Type", "application/json")
+	res, err := s.client.Do(req)
+	if err != nil {
+		return false, nil, err
+	}
+	defer res.Body.Close()
+	body, err := io.ReadAll(res.Body)
+	if err != nil {
+		return false, nil, err
+	}
+
+	if res.StatusCode == http.StatusOK {
+		var tokenRes tokenResponse
+		if err := json.Unmarshal(body, &tokenRes); (err != nil || tokenRes == tokenResponse{}) {
```

Comment:
This check is necessary because `json.Unmarshal` doesn't actually return an error if the input doesn't match the struct. Comparing to an empty struct is a way to validate that unmarshal succeeded _and_ it actually unmarshalled data.

Reply:
Do you know of a way to generate PATs in this style for testing?

I do not know such a way. Now dockerhub generates dckr_ pattern. But I have seen one in the wild and it's valid and working. That's how my tests pass.

Reply:
@rosecodym can we please add this PR to main?